File: app/src/app/core/models/invitation.model.ts
export type InvitationStatus = 'ACTIVE' | 'EXPIRED' | 'USED' | 'REVOKED';

export interface Invitation {
  id: number;
  email: string;
  invitedByEmail: string;
  status: InvitationStatus;
  token: string | null;
  createdAt: string;
  expiresAt: string;
  usedAt: string | null;
  revokedAt: string | null;
}

export interface InvitationCreated {
  token: string;
  expiresAt: string;
}

export interface CreateInvitationRequest {
  email: string;
}

export interface AcceptInviteRequest {
  token: string;
  password: string;
  displayName: string;
  currency: string;
  timezone: string;
}

export interface InviteLookup {
  email: string;
}

File: flutter/lib/src/features/admin/data/invitation_model.dart
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

import 'package:freezed_annotation/freezed_annotation.dart';

part 'invitation_model.freezed.dart';
part 'invitation_model.g.dart';

enum InvitationStatus {
  @JsonValue('ACTIVE')
  active,
  @JsonValue('EXPIRED')
  expired,
  @JsonValue('USED')
  used,
  @JsonValue('REVOKED')
  revoked,
}

@freezed
class Invitation with _$Invitation {
  const factory Invitation({
    required int id,
    required String email,
    required String invitedByEmail,
    required InvitationStatus status,
    String? token,
    required DateTime createdAt,
    required DateTime expiresAt,
    DateTime? usedAt,
    DateTime? revokedAt,
  }) = _Invitation;

  factory Invitation.fromJson(Map<String, dynamic> json) =>
      _$InvitationFromJson(json);
}

@freezed
class InvitationCreated with _$InvitationCreated {
  const factory InvitationCreated({
    required String token,
    required DateTime expiresAt,
  }) = _InvitationCreated;

  factory InvitationCreated.fromJson(Map<String, dynamic> json) =>
      _$InvitationCreatedFromJson(json);
}
